<!doctype html>
<html lang="de">
<head>
	<meta charset="utf-8">
	<title>Unterschiedliche Error-Typen in JavaScript</title>
	<meta name="description" content="">
	<meta name="author" content="Adam Beres-Deak">
	
	<meta name="apple-mobile-web-app-capable" content="yes" />
	<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent" />
	<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">
	<link rel="stylesheet" href="css/reveal.min.css">
	<link rel="stylesheet" href="css/theme/beige.css">
	<link rel="stylesheet" href="lib/css/zenburn.css">
	<!-- If the query includes 'print-pdf', include the PDF print sheet -->
	<script>
		if( window.location.search.match( /print-pdf/gi ) ) {
			var link = document.createElement( 'link' );
			link.rel = 'stylesheet';
			link.href = 'css/print/pdf.css';
			document.getElementsByTagName( 'head' )[0].appendChild( link );
		}
	</script>
	<!--[if lt IE 9]><script src="lib/js/html5shiv.js"></script><![endif]-->
	<style>
		h1 { font-size: 60px !important; }
		/*.reveal .slides > section, .reveal .slides > section > section { padding: 0; };*/
		h2.noTextTransform { text-transform: none !important; }
		code, pre { text-transform: none !important; }
	</style>
</head>
<body>
	<script>
		/*(function() {
			function log(message, url, lineNumber) {
				console.log(message);
				//console.error(message, url, lineNumber);
			}

			if (window.addEventListener) { window.addEventListener('error', log); }
			else if (window.attachEvent) { window.attachEvent('onerror', log); }
			else {
				var onError = window.onerror;
				window.onerror = function() {
					log.apply(this, arguments);
					onError.apply(this, arguments);
				};
			}
		}());
		//throw new TypeError("sdfsdf");

		var T = function(message){
			this.name = 'T';
			this.message = message;
			//Error.apply(this, arguments);
		};

		T.prototype = new Error();
		T.constructor = T;

		//console.log('xxxx');
		throw new T("1");
		try {
			throw new T("fsfsf");
		} catch(ex) {
			//console.error(ex);
		}
		*/
	</script>
	<div class="reveal">
		<div class="slides">
			<section>
				<h1>Unterschiedliche Error-Typen in JavaScript</h1>
				<p>
					<a target="_blank" href="http://bdadam.com">Adam Beres-Deak</a> /
					<a target="_blank" href="http://bdadam.com">bdadam.com</a> /
					<a target="_blank" href="http://twitter.com/bdadamm">@bdadamm</a>
				</p>
			</section>
			<section>
				Über mich
			</section>
			<section data-markdown>
				## Agenda

				1. Fehlerbehandlung in JavaScript
				2. Eingebaute und benutzerdefinierte Error-Typen
				3. Fehlerbehandlung in asynchron Funktionen
				4. Fehlerlogging
			</section>

			<section data-markdown>
				## 1. try/catch/finally/throw

				```JavaScript
				try {
					throw new Error('Something bad happened.');
				}
				catch(ex) {
					console.log(ex.name); // Error
					console.log(ex.message); // Something bad happened.
					console.log(ex.stack); // Error: Something bad happened.    at ...
				}
				finally {
					console.log('Finally always wins.');
				}

				```
			</section>
			<section data-markdown>
				## 3 mögliche Formen:
				* try - catch
				* try - finally
				* try - catch - finally
			</section>
			<section data-markdown>
				## Das `throw` Statement

				* Es wirft eine Exception
				* Die Ausführung der aktuellen Funktion wird beendet,
				* und springt zum nächsten `catch` Block.
				* Falls es kein `catch` gibt, das Program wird beendet.
			</section>
			<section data-markdown>
				## Finally always wins!

				```JavaScript
				function trueOrFalse() {
					try {
						throw new Error('An error occured.');
					}
					catch(ex) {
						return false;
					}
					finally {
						return true;
					}
				}

				console.log(trueOrFalse()); // ????
				```
			</section>
			<section data-markdown>
				## Finally always wins! /2

				Der Code in `finally` Block läuft auch wenn:

				* `return` in `try` Block
				* Code in `catch` Block wirft eine Exception
				* kein `catch` Block
			</section>

			<section data-markdown>
				## 2. Eingebaute und benutzerdefinierte Error-Typen
			</section>
			<section data-markdown>
				## Natives Error Objekt

				```JavaScript
				throw new Error('Error message');
				```
			</section>
			<section data-markdown>
				## Exceptions sind Freunde nicht Feinde

				* Exceptions erzählen was passiert ist
				* Wo ist es schief gelaufen
				* Sie helfen beim Debuggen
			</section>
			<section data-markdown>
				## Man sollte sinnvolle und sprechende Exceptions werfen

				zumindest das `message` Property einer Ausnahme setzen
			</section>
			<section data-markdown>
				```JavaScript
				function throwsAnException() {
					throw new Error('Something bad happened.');
				}

				try {
					throwsAnException();
				}
				catch(ex) {
					console.log(error.name); // Error
					console.log(error.message); // Something bad happened.
					console.log(ex.stack);
				}
				```
			</section>
			<section data-markdown>
				## Stack (ausser Internet Explorer)

				```
				Error: Something bad happened.
				    at throwsAnException (http://127.0.0.1:3000/:184:12)
				    at http://127.0.0.1:3000/:188:6
				    at http://127.0.0.1:3000/:193:5 (index):191
				```
			</section>
			<section data-markdown>
				## Was kann man noch besser machen?
			</section>
			<section>
				<section><h2>Einen von den nativen Error Typen verwenden</h2></section>
				<section><h2>Welche sind diese?</h2></section>
				<section><h2 class="noTextTransform">Error</h2></section>
				<section><h2 class="noTextTransform">TypeError</h2></section>
				<section><h2 class="noTextTransform">EvalError</h2></section>
				<section><h2 class="noTextTransform">URIError</h2></section>
				<section><h2 class="noTextTransform">SyntaxError</h2></section>
				<section><h2 class="noTextTransform">RangeError</h2></section>
				<section><h2 class="noTextTransform">ReferenceError</h2></section>
			</section>
			<section><h2>Benutzerdefinierte Error Types</h2></section>
			<section data-markdown>
				```JavaScript
				function MyError(message) {
					this.message = message;
					this.stack = (new Error()).stack;
				}

				MyError.prototype = new Error();
				MyError.constructor = MyError;
				MyError.prototype.name = 'MyError';

				try {
					throw new MyError('Error message.');
				} catch(ex) {
					console.log(ex instanceof Error); // true
					console.log(ex instanceof MyError); // true
				}
				```
			</section>
			<section data-markdown>
				## Unterschiedliche Behandlung je nach Error Typ

				z.B. in C#:

				```c#
				try {...}
				catch (ArgumentException ex) { ... }
				catch (FileNotFoundException ex) { ... }
				```
			</section>
			<section data-markdown>
				## Der Standard Weg

				```JavaScript
				try { ... }
				catch(ex) {
					if (ex instanceof URIError) { ... }
					else if (ex instanceof TypeError) { ... }
					else if (ex instanceof MyCustomError) { ... }
				}
				```
			</section>
			<section data-markdown>
				## Der Firefox Weg

				```JavaScript
				try {
					myroutine(); // may throw some exceptions
				}
				catch (e if e instanceof TypeError) { ... }
				catch (e if e instanceof RangeError) { ... }
				catch (e) {
					logMyErrors(e); // pass exception object to error handler
				}
				```

				[mehr Infos auf MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/try...catch)
			</section>

			<section><h2>3. Fehlerbehandlung in asynchron Funktionen</h2></section>
			<section data-markdown>
				## Mit callback Funktionen

				```JavaScript
				function doSomethingAsync(onSuccess, onError) {
					$.ajax(url, {
						success: onSuccess,
						error: onError
					});
				}

				doSomethingAsync(
					function() { console.log('success'); },
					function() { console.log('error'); }
				);
				```
			</section>
			<section data-markdown>
				## Wie in Node.js

				eine callback Funktion

				```JavaScript
				fs.readFile('data.txt', function(error, result) {
					if (error) {
						// error handling
					}
				});
				```
			</section>
			<section data-markdown>
				## Wie in Node.js /2

				```JavaScript
				function doSomethingAsync(callback) {
					$.ajax(url, {
						success: function(data) { callback(null, data); },
						error: function() { callback(new Error('Network error.')); }
					});
				}
				```
			</section>
			<section data-markdown>
				## Promises
				
				```JavaScript
				doSomethingAsync()
					.then(onFulfilled, onRejected)
					.catch(function() { ... });
				```
			</section>
			<section data-markdown>
				## Promises /2

				```JavaScript
				function doSomethingAsync() {
					return new Promise(function(resolve, reject) {
						$.ajax(url, {
							success: function(data) { resolve(data); },
							error: function() { reject(new Error('Network error')); }
						});
					});
				}
				```
			</section>

			<section><h2>4. Error Logging</h2></section>
			<section data-markdown>
				```JavaScript
				(function() {
					function log(message, url, lineNumber) {
						// the actual logging happens here
					}

					if (window.addEventListener) { window.addEventListener('error', log); }
					else if (window.attachEvent) { window.attachEvent('onerror', log); }
					else {
						var onError = window.onerror;
						window.onerror = function() {
							log.apply(arguments);
							onError.apply(arguments);
						};
					}
				}());
				```
			</section>
			<section><h2>Wie kann man den Fehler loggen?</h2></section>
			<section data-markdown>
				## Google Analytics

				```JavaScript
				function log(message, url, lineNumber) {
					
				}
				```
			</section>
			<section data-markdown>
				## Ajax-Request

				```JavaScript
				function log(message, url, lineNumber) {
					$.ajax(...)
					// oder
					var xmlhttp=new XMLHttpRequest();
					xmlhttp.open('GET', '/url/to/error/logger', true);
					xmlhttp.send();
				}
				```
			</section>
			<section data-markdown>
				## Image Request

				```JavaScript
				function log(message, url, lineNumber) {
					new Image('/url/to/error/logger');
				}
				```
			</section>
		</div>
	</div>
	<script src="lib/js/head.min.js"></script>
	<script src="js/reveal.min.js"></script>
	<script>
		// Full list of configuration options available here:
		// https://github.com/hakimel/reveal.js#configuration
		Reveal.initialize({
			controls: true,
			progress: true,
			history: true,
			center: true,
			slideNumber: true,
			mouseWheel: true,

			theme: Reveal.getQueryHash().theme,
			transition: Reveal.getQueryHash().transition || 'linear', // default/cube/page/concave/zoom/linear/fade/none
			dependencies: [
				{ src: 'lib/js/classList.js', condition: function() { return !document.body.classList; } },
				{ src: 'plugin/markdown/marked.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
				{ src: 'plugin/markdown/markdown.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
				{ src: 'plugin/highlight/highlight.js', async: true, callback: function() { hljs.initHighlightingOnLoad(); } },
				{ src: 'plugin/zoom-js/zoom.js', async: true, condition: function() { return !!document.body.classList; } },
				{ src: 'plugin/notes/notes.js', async: true, condition: function() { return !!document.body.classList; } }
			]
		});

		window.addEventListener('mousewheel', handleScroll);
		window.addEventListener('DOMMouseScroll', handleScroll);

		function handleScroll(e) {
			e.preventDefault();
			var delta = e.detail || -e.wheelDelta;

			if (delta < 0) { Reveal.prev(); }
			else { Reveal.next(); }
		}
	</script>
</body>
</html>
